src/componententity.cpp

```diff
@@ -262,18 +375,36 @@ const ComponentPtr& ComponentEntity::getComponent(size_t index) const
     return mPimpl->mComponents.at(index);
 }
 
-ComponentPtr ComponentEntity::getComponent(const std::string &name)
+ComponentPtr ComponentEntity::getComponent(const std::string &name, bool searchEncapsulated)
 {
-    auto result = mPimpl->findComponent(name);
-    size_t index = result - mPimpl->mComponents.begin();
-    return mPimpl->mComponents.at(index);
+    ComponentPtr foundComponent = nullptr;
+    if (containsComponentInThis(name)) {
+        foundComponent = getComponentInThis(name);
+    } else if (searchEncapsulated) {
+        for (size_t i = 0; i < componentCount(); ++i) {
+            foundComponent = getComponent(i)->getComponent(name, searchEncapsulated);
+            if (foundComponent) {
+                break;
+            }
+        }
+    }
+    return foundComponent;
 }
 
-const ComponentPtr& ComponentEntity::getComponent(const std::string &name) const
+const ComponentPtr ComponentEntity::getComponent(const std::string &name, bool searchEncapsulated) const
 {
-    auto result = mPimpl->findComponent(name);
-    size_t index = result - mPimpl->mComponents.begin();
-    return mPimpl->mComponents.at(index);
+    if (containsComponent(name, searchEncapsulated)) {
+        if (containsComponentInThis(name)) {
+            return getComponentInThis(name);
+        } else if (searchEncapsulated) {
+            for (size_t i = 0; i < componentCount(); ++i) {
+                if (getComponent(i)->containsComponent(name, searchEncapsulated)) {
+                    return getComponent(i)->getComponent(name, searchEncapsulated);
+                }
+            }
+        }
+    }
+    throw std::out_of_range("Named component not found.");
```

Comment:
Just to update, I'm having trouble getting the Windows build to throw this exception at the moment. I should be able to interactively debug this on Monday through RDP but will work on the validation side in the meantime.